src/python/pants/backend/go/tailor.py

```diff
@@ -0,0 +1,144 @@
+    PutativeTargetsRequest,
+    group_by_dir,
+)
+from pants.engine.fs import PathGlobs, Paths
+from pants.engine.internals.selectors import Get
+from pants.engine.rules import collect_rules, rule
+from pants.engine.target import UnexpandedTargets
+from pants.engine.unions import UnionRule
+from pants.util.logging import LogLevel
+
+
+@dataclass(frozen=True)
+class PutativeGoPackageTargetsRequest(PutativeTargetsRequest):
+    pass
+
+
+@rule(level=LogLevel.DEBUG, desc="Determine candidate Go `go_package` targets to create")
+async def find_putative_go_package_targets(
+    request: PutativeGoPackageTargetsRequest, all_owned_sources: AllOwnedSources
+) -> PutativeTargets:
+    all_go_files = await Get(Paths, PathGlobs, request.search_paths.path_globs("*.go"))
+    unowned_go_files = set(all_go_files.files) - set(all_owned_sources)
+
+    putative_targets = []
+    for dirname, filenames in group_by_dir(unowned_go_files).items():
+        putative_targets.append(
+            PutativeTarget.for_target_type(
+                GoPackage,
+                dirname,
+                os.path.basename(dirname),
+                sorted(filenames),
+            )
+        )
+
+    return PutativeTargets(putative_targets)
+
+
+@dataclass(frozen=True)
+class PutativeGoModuleTargetsRequest(PutativeTargetsRequest):
+    pass
+
+
+@rule(level=LogLevel.DEBUG, desc="Determine candidate Go `go_module` targets to create")
+async def find_putative_go_module_targets(
+    request: PutativeGoModuleTargetsRequest, all_owned_sources: AllOwnedSources
+) -> PutativeTargets:
+    all_go_mod_files = await Get(Paths, PathGlobs, request.search_paths.path_globs("go.mod"))
+    unowned_go_mod_files = set(all_go_mod_files.files) - set(all_owned_sources)
+
+    putative_targets = []
+    for dirname, filenames in group_by_dir(unowned_go_mod_files).items():
+        putative_targets.append(
+            PutativeTarget.for_target_type(
+                GoModule,
+                dirname,
+                os.path.basename(dirname),
+                sorted(filenames),
+            )
+        )
+
+    return PutativeTargets(putative_targets)
+
+
+def compute_go_external_module_target_name(name: str, version: str) -> str:
+    return f"{name.replace('/', '_')}_{version}"
+
+
+@dataclass(frozen=True)
+class PutativeGoExternalModuleTargetsRequest(PutativeTargetsRequest):
+    pass
+
+
+@rule(level=LogLevel.DEBUG, desc="Determine candidate Go `go_external_module` targets to create")
+async def find_putative_go_external_module_targets(
+    request: PutativeGoExternalModuleTargetsRequest, _all_owned_sources: AllOwnedSources
+) -> PutativeTargets:
+    # Unlike ordinary tailor invocations, this rule looks at existing `go_module` targets and not at actual
+    # source files because it infers `go_external_module` targets based on go.mod contents. (This may require
+    # invoking `tailor` first to create `go_module` targets and then again to create `go_external_module`
+    # targets.)
```

Comment:
Would you mind referring to https://github.com/pantsbuild/pants/issues/7022 here with a TODO? It seems like this would pretty clearly be a macro if macros were able to invoke processes.

Reply:
Done.